src/app/thumbnails.h

```diff
@@ -0,0 +1,154 @@
+// Aseprite
+// Copyright (C) 2016  Carlo "zED" Caputo
+//
+// This program is free software; you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 2 as
+// published by the Free Software Foundation.
+
+#ifndef APP_THUMBNAILS_H_INCLUDED
+#define APP_THUMBNAILS_H_INCLUDED
+#pragma once
+
+#include "base/shared_ptr.h"
+#include "base/connection.h"
+#include "she/surface.h"
+#include "doc/object_id.h"
+#include "doc/object.h"
+#include "doc/frame.h"
+
+#include <map>
+#include <set>
+#include <utility>
+
+namespace doc {
+  class Cel;
+  class Image;
+  class Document;
+}
+
+namespace app {
+  class Document;
+
+  namespace thumb {
+
+    typedef long long Sequence;
+
+    typedef std::pair< int, int > Dimension;
+
+    class Request {
+    public:
+      Request();
+      Request(const app::Document* doc, const doc::frame_t frm, const doc::Image* img, const Dimension dim);
+      const Dimension dimension;
+      const app::Document* document;
+      const doc::Image* image;
+      const doc::frame_t frame;
+      const Sequence timestamp;
+      bool updated;
+    private:
+      static Sequence m_sequence;
+    };
+
+    class Tag {
+    public:
+      Tag();
+      Tag(const Request& req);
+      Dimension dimension;
+      doc::ObjectId document;
+      doc::ObjectId image;
+      Sequence timestamp;
+    };
+
+    struct TagRefComp
+    {
+      bool operator()(const Tag* lhs, const Tag* rhs) const;
+    };
+    typedef std::set< Tag*, TagRefComp > RecentlyUsed;
+
+    class Tier {
```

Comment:
It would be nice if you can comment in the issue about your design decisions. At a quick glance, I don't understand why there is a full hierarchy of classes here (`Tier`, `Data`, `Directory`, `Surface`, `Image`, `Document`, `Cache`). Also try to use other class names, I know that `thumb` namespace avoid name collisions, but it make harder to read the code. (And example of something we're dealing with: `doc::Document` vs `app::Document`, something I've to refactor because it's a pain to deal with both classes.)